### PR TITLE
Adding IfcVoxelGrid

### LINKED ISSUE #117: [IFC Tunnel] Voxel geometric representation and data management
A requirement from IFC Tunnel stakeholders is to have a representation based on voxels. Each voxel has to provide the ability to carry a physical characteristic in the form of a value or a directed value i.e. vector.

[DOCUMENTATION]
[Voxel representation.pdf](https://github.com/bSI-InfraRoom/IFC-Specification/files/6863942/Voxel.representation.pdf)

Initial feedback and derived issues:
#194 
#195 
#202 
TODO ... Adding some more feedback

### PR DESCRIPTION
**Pull request by SergejMuhic**
Resolves #117 

Added:
1. `IfcVoxelGrid` definition

### CODE REVIEW

**Comment by anborr** [comment]
Looks good. 
The only concern I have is whether `NumberOfVoxelsX `would be a better fit than `GridSizeX`.
There is a typo in the documentation of the `Voxels` attribute: "...array is one dimentional" must be become "...array is one-dimensional...".
Issue #198 was not considered, i.e. we cannot represent hierachies of Voxel Grids. This would require to have a `Child` relationship to 1..n `IfcVoxelGrid `objects
I propose to add a figure to the documentation.

**Review by anborr — CHANGES_REQUESTED** [review]
Looks good.
The only concern I have is whether `NumberOfVoxelsX `would be a better fit than `GridSizeX`.
There is a typo in the documentation of the `Voxels` attribute: "...array is one dimentional" must be become "...array is one-dimensional...".
Issue #198 was not considered, i.e. we cannot represent hierachies of Voxel Grids. This would require to have a `Child` relationship to 1..n `IfcVoxelGrid` objects
I propose to add a figure to the documentation.

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+			<Documentation>Size of the voxel grid in the X axis.</Documentation>
+		</DocAttribute>
+		<DocAttribute Name="GridSizeY" UniqueId="70c30c22-0e2a-446d-8a02-31355e75c4e7" DefinedType="IfcInteger" AttributeFlags="1">
+			<Documentation>Size of the voxel grid in the Y axis. If not given, the value from _GridSizeX_ shall be taken.</Documentation>
+		</DocAttribute>
+		<DocAttribute Name="GridSizeZ" UniqueId="bbb9efce-f6f6-4834-a122-fbef29b024bd" DefinedType="IfcInteger" AttributeFlags="1">
+			<Documentation>Size of the voxel grid in the Z axis. If not given, the value from _GridSizeX_ shall be taken.</Documentation>
+		</DocAttribute>
+		<DocAttribute Name="Voxels" UniqueId="14ee4900-2fe3-4e22-9af9-7e201ba42225" DefinedType="IfcBoolean" AggregationType="1" AggregationLower="1" AggregationUpper="0">
+			<Documentation>Indication of voxels on the grid. The array is one-dimentional where values are distributed in the following order:
```
Comment (SergejMuhic) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
			<Documentation>Indication of voxels on the grid. The array is one-dimensional where values are distributed in the following order:
```

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+		<DocAttribute Name="VoxelSizeX" UniqueId="ceb76c5f-caa2-4d51-98ab-52b1142cb762" DefinedType="IfcNonNegativeLengthMeasure">
+			<Documentation>Size of voxels in the X axis.</Documentation>
+		</DocAttribute>
+		<DocAttribute Name="VoxelSizeY" UniqueId="781db54c-ab64-4392-afd7-7d327e4f106a" DefinedType="IfcNonNegativeLengthMeasure" AttributeFlags="1">
+			<Documentation>Size of voxels in the Y axis. If not given, the value from _VoxelSizeX_ shall be taken.</Documentation>
+		</DocAttribute>
+		<DocAttribute Name="VoxelSizeZ" UniqueId="946806c3-1ace-4539-b0a5-5c2d882b98c8" DefinedType="IfcNonNegativeLengthMeasure" AttributeFlags="1">
+			<Documentation>Size of voxels in the Z axis. If not given, the value from _VoxelSizeX_ shall be taken.</Documentation>
+		</DocAttribute>
+		<DocAttribute Name="GridSizeX" UniqueId="1062135e-7018-4ec4-9ef7-3ca487268280" DefinedType="IfcInteger">
```
Comment (larswik) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
		<DocAttribute Name="NumberOfVoxelsX" UniqueId="1062135e-7018-4ec4-9ef7-3ca487268280" DefinedType="IfcInteger">
```

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+			<Documentation>Size of the voxel grid in the X axis.</Documentation>
```
Comment (larswik) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
			<Documentation>Number of voxels along the X axis.</Documentation>
```

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+		</DocAttribute>
+		<DocAttribute Name="GridSizeY" UniqueId="70c30c22-0e2a-446d-8a02-31355e75c4e7" DefinedType="IfcInteger" AttributeFlags="1">
```
Comment (larswik) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
		<DocAttribute Name="NumberOfVoxelsY" UniqueId="70c30c22-0e2a-446d-8a02-31355e75c4e7" DefinedType="IfcInteger" AttributeFlags="1">
```

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+			<Documentation>Size of the voxel grid in the Y axis. If not given, the value from _GridSizeX_ shall be taken.</Documentation>
```
Comment (larswik) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
			<Documentation>Number of voxels along the Y axis. If not given, the value from _NumberOfVoxelsX_ shall be taken.</Documentation>
```

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		</DocAttribute>
+		<DocAttribute Name="GridSizeZ" UniqueId="bbb9efce-f6f6-4834-a122-fbef29b024bd" DefinedType="IfcInteger" AttributeFlags="1">
```
Comment (larswik) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
		<DocAttribute Name="NumberOfVoxelsZ" UniqueId="bbb9efce-f6f6-4834-a122-fbef29b024bd" DefinedType="IfcInteger" AttributeFlags="1">
```

File: IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+			<Documentation>Size of the voxel grid in the Z axis. If not given, the value from _GridSizeX_ shall be taken.</Documentation>
```
Comment (larswik) [inline IFC4x3/Sections/Resource definition data schemas/Schemas/IfcGeometricModelResource/Entities/IfcVoxelGrid/DocEntity.xml]:
```suggestion
			<Documentation>NumberOf voxels along the Z axis. If not given, the value from _NumberOfVoxelsX_ shall be taken.</Documentation>
```

**Review by larswik — CHANGES_REQUESTED** [review]

**Review by larswik — APPROVED** [review]

**Pull request merged**